### ISSUE TITLE
No audio in ZTE NX629J

### ISSUE THREAD

**Issue by fzurita** [google/oboe#1121]
Android version(s): 9.0
Android device(s): ZTE NX629J
Oboe version: 1.4.3
App name used for testing: M64Plus FZ
(Please try to reproduce the issue using the OboeTester or an Oboe sample.)
 
**Short description**
I'm getting complaints of some devices having no audio after switching to the OBOE library.

**Steps to reproduce**
Play a game in the app

**Expected behavior**
Oboe should produce audio output

**Actual behavior**
Stream fails to start for an unknown reason.

**Device**
ZTE NX629J at the very least. Probably other "red magic" devices as well.

I don't have access to the device, but it's using a qualcomm SD 855, arm64-v8a, and it reports as supporting aaudio

```
for p in \
    ro.product.brand ro.product.manufacturer ro.product.model \
    ro.product.device ro.product.cpu.abi ro.build.description \
    ro.hardware ro.hardware.chipname ro.arch "| grep aaudio";
    do echo "$p = $(adb shell getprop $p)"; done
```

**Any additional context**

Unfortunately, I don't have access to the specific device, so the best I can do is report that the stream fails to start for specific devices.

If it helps, the application is open source. The audio handling can be found here: https://github.com/mupen64plus-ae/mupen64plus-ae/blob/master/mupen64plus-audio-android/src/AudioHandler.cpp

Also, if you can pinpoint anything that could be improved in the usage of the library, that wouldn't be a bad thing either.

Another data point, I have done some Google searching online and it seems that users complain about no audio in other games as well on those specific devices.

**Comment by fzurita** [comment]
I was able to obtain a logcat from the device:
[oboe_logcat.txt](https://github.com/google/oboe/files/5612909/oboe_logcat.txt)

First, through testing through the user, who is being very helpful, I was able to find that forcing the audio to SLES through the builder ```setAudioApi``` allows sound to work.

I also updated to 1.5 and the same issue is still present.

When opening the stream through the builder, the result shows up as "OK". So I can't fallback to SLES on failure to open the stream.

One add thing I noticed is this from the AudioStreamInternal_Client logs:
 ``setBufferSize() req = 192 => 192```

But ```getFramesPerBurst``` is returning 96.

**Comment by philburk** [comment]
Thanks for the logcat. I can see logs from AudioStreamInternalPlay_Client, which means that the device opened an AAudio MMAP stream. But MMAP is apparently not implemented correctly on that device. I will work with the manufacturers to investigate this further.  THANK YOU so much for reporting this.

**Comment by philburk** [comment]
> setBufferSize() req = 192 => 192
> But getFramesPerBurst is returning 96

Yes, that means that we set the buffer size to two bursts. It's OK. Some people call that a "double buffered buffer". But I find that confusing.

**Comment by fzurita** [comment]
I see, good to know that that's fine.

For now, I published an update that checks if the manufacturer string of a device is "nubia", and if that's the case then my app will force SLES audio instead of AAudio, which seems to work ok for now. Maybe I can remove it once the library gets updated.

**Comment by philburk** [comment]
>  checks if the manufacturer string of a device is "nubia",

@fzurita - I would like to add a similar workaround to Oboe based on system properties. So I will need some exact information.  Can you provide the name and value of the system property you used? Or can I get a dump of properties from that device?

    adb shell getprop > propdump.txt

and then attach the file.

**Comment by fzurita** [comment]
Unfortunately, I don't have access to the device and the user is not handy with adb, he was using the built in log reporting tool within the application itself.

I'm using this property:
```android.os.Build.MANUFACTURER```

This is basically my check:
```android.os.Build.MANUFACTURER.toLowerCase().contains("nubia")```

https://github.com/mupen64plus-ae/mupen64plus-ae/commit/39b3e9d63810be2882388f527bcae4f81b57289c

**Comment by fzurita** [comment]
On a side note, I'm getting reports of no audio when screen sharing apps are used. I was under the impression that the library is supposed to go to a "shared" mode when "exclusive" is requested but it can't be obtained?

**Comment by philburk** [comment]
> On a side note, I'm getting reports of no audio when screen recording apps are used.

Is that related to this ZTE problem? If not then please file a separate Issue.

**Comment by fzurita** [comment]
It's not related, I'll make a separate issue once I find out more information and I know it's not my problem.

**Comment by philburk** [comment]
This issue is being tracked internally at Android as:
b/174501337 | AAudio MMAP produces no sound from speaker on a ZTE NX629J

**Comment by dturner** [comment]
AI: If it's a `nubia` device we will not enable MMAP streams.

**Comment by philburk** [comment]
Qualcomm says they added MMAP support for the SM8150 in a QC update. It is not available on Pie (9.0).
So we should add a Quirk to not use MMAP on SM8150 before Q.

**Comment by philburk** [comment]
The SM8150 is used in Pixel 4 and 4a so we can use that for vierying detection of the SM8150.
The Pixel 4 and 4A were released with Q.
Here is a [list of phones that use the SM8150](https://phonesdata.com/en/chipset/qualcomm/sm8150-snapdragon-855-(7-nm)/).

From Pixel 4:
ro.soc.manufacturer=Qualcomm
ro.soc.model=SM8150
